src/librustc_metadata/decoder.rs

```diff
@@ -496,30 +467,106 @@ impl<'a, 'tcx> CrateMetadata {
         }
     }
 
-    pub fn item_name(&self, item_index: DefIndex) -> Symbol {
-        self.def_key(item_index)
-            .disambiguated_data
-            .data
-            .get_opt_name()
-            .expect("no name in item_name")
-            .as_symbol()
+    pub fn item_name(&self, item_index: DefIndex, sess: &Session) -> Symbol {
+        if !self.is_proc_macro(item_index, sess) {
+            self.def_key(item_index, sess)
+                .disambiguated_data
+                .data
+                .get_opt_name()
+                .expect("no name in item_name")
+                .as_symbol()
+        } else {
+            self.get_proc_macros(sess).unwrap()[&item_index].name
+        }
     }
 
-    pub fn def_kind(&self, index: DefIndex) -> Option<DefKind> {
-        if !self.is_proc_macro(index) {
+    pub fn def_kind(&self, index: DefIndex, sess: &Session) -> Option<DefKind> {
+        if !self.is_proc_macro(index, sess) {
             self.entry(index).kind.def_kind()
         } else {
             Some(DefKind::Macro(
-                self.proc_macros.as_ref().unwrap()[index.to_proc_macro_index()].1.macro_kind()
+                self.get_proc_macros(sess).unwrap()[&index].ext.macro_kind()
             ))
         }
     }
 
     pub fn get_span(&self, index: DefIndex, sess: &Session) -> Span {
-        match self.is_proc_macro(index) {
-            true => DUMMY_SP,
-            false => self.entry(index).span.decode((self, sess)),
+        self.entry(index).span.decode((self, sess))
+    }
+
+
+    pub fn get_proc_macros(&self, sess: &Session) -> Option<FxHashMap<DefIndex, FullProcMacro>>{
+        let proc_macros = self.root.proc_macro_data.as_ref().map(|proc_macros| {
+            let proc_macros = proc_macros.decode((self, sess));
```

Comment:
Wait, you don't even need `sess` to do this, if you just have a `DefIndex`!

Reply:
I could change this `decode` call to just use `self`. However, we still need `sess` in order to check `self.sess.opts.debugging_opts.dual_proc_macros`, and to be able to call `get_span` in `load_proc_macros`

Reply:
Right, but to check whether something is a proc macro `sess` shouldn't be needed.